From edc12b77eb66c20ab43552b4ffa298659f25fe49 Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Fri, 22 May 2009 14:28:37 +0800
Subject: [PATCH] net/ucc_geth: fix set mac address bug in ucc_geth

Currently we map the ucc register in the ucc_geth_open fuction.
This will cause core dump if we set mac address before net device
is opened. So don't set mac addr registers if device is not running.

If the device is running, we also need a lock when setting the mac
address or else we will get crashed packets.

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ucc_geth.c |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)

diff --git a/drivers/net/ucc_geth.c b/drivers/net/ucc_geth.c
index fee9fcb..e07b11d 100644
--- a/drivers/net/ucc_geth.c
+++ b/drivers/net/ucc_geth.c
@@ -3688,11 +3688,21 @@ static int ucc_geth_set_mac_address(struct net_device *dev, void *p)
 {
 	struct ucc_geth_private *ugeth = netdev_priv(dev);
 	struct sockaddr *addr = p;
+	unsigned long flags;
 
 	if (!is_valid_ether_addr(addr->sa_data))
 		return -EADDRNOTAVAIL;
 
 	memcpy(dev->dev_addr, addr->sa_data, dev->addr_len);
+
+	/*
+	 * If device is not running, we will set mac addr register
+	 * when opening the device.
+	 */
+	if (!netif_running(dev))
+		return 0;
+
+	spin_lock_irqsave(&ugeth->lock, flags);
 	init_mac_station_addr_regs(dev->dev_addr[0],
 				   dev->dev_addr[1],
 				   dev->dev_addr[2],
@@ -3701,6 +3711,7 @@ static int ucc_geth_set_mac_address(struct net_device *dev, void *p)
 				   dev->dev_addr[5],
 				   &ugeth->ug_regs->macstnaddr1,
 				   &ugeth->ug_regs->macstnaddr2);
+	spin_unlock_irqrestore(&ugeth->lock, flags);
 
 	return 0;
 }
-- 
1.6.0.4

